src/transformers/models/gpt_neo/modeling_gpt_neo.py

```diff
@@ -130,7 +130,151 @@ def load_tf_weights_in_gpt_neo(model, config, gpt_neo_checkpoint_path):
+            :obj:`True` else a tensor of shape :obj:`[batch_size, window_size + block_length, num_blocks, ...]`
+        """
+        if len(tensor.shape) == 3:
+            padding_side = (0, 0, window_size, 0)
+        elif len(tensor.shape) == 2:
+            padding_side = (window_size, 0)
+        else:
+            raise ValueError(f"Input tensor rank should be one of [2, 3], but is: {len(tensor.shape)}")
+
+        padded_tensor = F.pad(tensor, padding_side, value=pad_value)
+        padded_tensor = padded_tensor.unfold(dimension=1, size=window_size + block_length, step=block_length)
+
+        if is_key_value:
+            padded_tensor = padded_tensor.transpose(-2, -1)
+        return padded_tensor
+
+    def _split_heads(self, tensor, num_heads, attn_head_size):
+        """
+        Splits hidden_size dim into attn_head_size and num_heads
+        """
+        new_shape = tensor.size()[:-1] + (num_heads, attn_head_size)
+        tensor = tensor.view(*new_shape)
+        if len(tensor.shape) == 5:
+            return tensor.permute(0, 1, 3, 2, 4)  # (batch, blocks, head, block_length, head_features)
+        elif len(tensor.shape) == 4:
+            return tensor.permute(0, 2, 1, 3)  # (batch, head, seq_length, head_features)
+        else:
+            raise ValueError(f"Input tensor rank should be one of [4, 5], but is: {len(tensor.shape)}")
+
+    def _merge_heads(self, tensor, num_heads, attn_head_size):
+        """
+        Merges attn_head_size dim and num_attn_heads dim into hidden_size
+        """
+        if len(tensor.shape) == 5:
+            tensor = tensor.permute(0, 1, 3, 2, 4).contiguous()
+        elif len(tensor.shape) == 4:
+            tensor = tensor.permute(0, 2, 1, 3).contiguous()
+        else:
+            raise ValueError(f"Input tensor rank should be one of [4, 5], but is: {len(tensor.shape)}")
+        new_shape = tensor.size()[:-2] + (num_heads * attn_head_size,)
+        return tensor.view(new_shape)
+
+    def _split_seq_length_dim_to(self, tensors, dim_factor_1, dim_factor_2, hidden_size):
+        """
+        Splits sequence length dim of tensors into `dim_factor_1` and `dim_factor_2` dims
+        """
+        batch_size = tensors.shape[0]
+        split_dim_shape = (batch_size, dim_factor_1, dim_factor_2)
+
+        if len(tensors.shape) == 3:
+            return torch.reshape(tensors, split_dim_shape + (hidden_size,))
+        elif len(tensors.shape) == 2:
+            return torch.reshape(tensors, split_dim_shape)
+        else:
+            raise ValueError(f"Input vector rank should be one of [2, 3], but is: {len(tensors.shape)}")
+
+    def _attn(self, query, key, value, causal_mask, masked_bias, attn_dropout, attention_mask=None, head_mask=None):
+        # Keep the attention weights computation in fp32 to avoid overflow issues
+        query = query.to(torch.float32)
+        key = key.to(torch.float32)
```

Comment:
Does this come from the original codebase?

Reply:
yes, see https://github.com/tensorflow/mesh/blob/master/mesh_tensorflow/transformer/attention.py#L74

Reply:
Hm, it looks like this is a new addition to mesh-tensorflow as of dec-2020 which they auto-default to true (typical >.>) https://github.com/tensorflow/mesh/commit/c2142a3b4b5f5eaf37a926d30525d2cf8334c65b#diff-c286eb8cfe654942200043ccf0baec37d48a8dc7642896f21ca425c1a8159e0c

I'm 95% certain this isn't in the mtf version we used to train gpt-neo

Reply:
I see. Thanks a lot for pointing this out. Let me run few training jobs without this to verify that it does not overflow in fp16, if yes, we can probably remove this.